### PR TITLE
Remove all-to-all operation from TSP/CP in Attention Segmentation Mask

### PR DESCRIPTION
**Pull request by dinodeep**
# Description

When performing TSP/CP with packed sequences, the segmentation mask produced for attention from ids could introduce a large all-to-all operation (transporting bytes on the order of `O(S^2)`). This PR avoids that by introducing an all-gather on the small non-broadcasted to reduce communication.

More specifically, this line
```
mask = decoder_segment_ids[:, :, None] == decoder_key_segment_ids[:, None, :]
```
would produce an all-to-all on the broadcasted `decoder_key_segment_ids[:, None, :]` when running with TSP/CP since `decoder_segment_ids` is `[B, S]` following the input data sharding. The sequence dimension is sharded under TSP/CP constraints; however, this introduces a large all-to-all collective when constructing the `mask`. 

To fix this, we add a sharding constraint on the right hand side of the equation. This all-gathers `decoder_key_segment_ids` on the sequence dimension for the right-hand side prior to broadcasting for a much smaller collective with the same sharded output. This reduces communication latency and can improve performance when training with large sequence lengths.

# Tests

I observed all-to-all operations (300-400us) in the HLO for a Llama3-70b training run executing with tensor_sequence_parallelism. Then, after running the same configuration but with the new changes, I did not observe an all-to-all instruction in the HLO, and instead, the operation was replaced with a small all-gather (30us latency).

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/attention_op.py [hunk 1]
```diff
@@ -752,7 +748,15 @@ def generate_attention_mask(
     if model_mode == MODEL_MODE_AUTOREGRESSIVE and decoder_segment_ids is not None:
       mask = decoder_segment_ids[:, None, None, None, :] == DECODING_ACTIVE_SEQUENCE_INDICATOR
     elif decoder_segment_ids is not None:
-      mask = decoder_segment_ids[:, :, None] == decoder_segment_ids[:, None, :]
+
+      # With TSP/CP, all-gather prior to broadcast to avoid large all-to-all on broadcasted ids.
+      if not self.config.using_pipeline_parallelism:
+        key_sharding = self._logical_to_mesh_axes((DATA_EMB_BATCH, None))
```
Comment (gobbleturk) [inline src/maxtext/layers/attention_op.py]: activation_embed_and logits is meant to only be used for the vocab activations (batch, sequence, vocab). You may need to create a different logical axis rule, or write a function explicitly all gather over TSP / CP

Comment (dinodeep) [reply src/maxtext/layers/attention_op.py]: Thanks for the review, I created a new logical axis rule specifically for the segment ids `"segment_ids_batch"`

File: src/maxtext/configs/base.yml [hunk 2]
```diff
@@ -507,8 +508,10 @@ mesh_axes: ['diloco', 'data', 'stage', 'fsdp', 'fsdp_transpose', 'context', 'con
 logical_axis_rules: [
                       ['circular_repeats', []],
                       # ==========================================
-                      # Vocabulary Embedding
+                      # Data / Vocabulary Embedding
                       # ==========================================
+                      # Data Segment IDs
+                      ['segment_ids_batch', ['data', 'stage', 'fsdp', 'fsdp_transpose', 'expert']],
```
Comment (NuojCheng) [inline src/maxtext/configs/base.yml]: could you move it under ` Standard MLP / Dense Layers / Model Structure` -> `Dense Activations`

Comment (dinodeep) [reply src/maxtext/configs/base.yml]: Moved under dense activations

**Review by NuojCheng — APPROVED** [review]

File: src/maxtext/configs/base.yml [hunk 3]
```diff
@@ -564,6 +565,7 @@ logical_axis_rules: [
                       # Standard MLP / Dense Layers / Model Structure
                       # ==========================================
                       # Dense Activations
+                      ['segment_ids_batch', ['data', 'stage', 'fsdp', 'fsdp_transpose', 'expert']],
```
Comment (gobbleturk) [inline src/maxtext/configs/base.yml]: can you remove stage from here? PP should not be sharding this

File: src/maxtext/layers/attention_op.py [hunk 4]
```diff
@@ -750,7 +748,15 @@ def generate_attention_mask(
     if model_mode == MODEL_MODE_AUTOREGRESSIVE and decoder_segment_ids is not None:
       mask = decoder_segment_ids[:, None, None, None, :] == DECODING_ACTIVE_SEQUENCE_INDICATOR
     elif decoder_segment_ids is not None:
-      mask = decoder_segment_ids[:, :, None] == decoder_segment_ids[:, None, :]
+
+      # With TSP/CP, all-gather prior to broadcast to avoid large all-to-all on broadcasted ids.
+      if not self.config.using_pipeline_parallelism:
```
Comment (gobbleturk) [inline src/maxtext/layers/attention_op.py]: I don't think PP is special here, should do the same all gather within microbatches of PP - should be more performant also for PP and simpler code

**Review by gobbleturk — APPROVED** [review]

**Review by huytransformer — APPROVED** [review]
I think might need a rebase since main has some new tests that could cause trouble
